src/value/element_stream_reader.rs

```diff
@@ -193,18 +203,11 @@ impl IonReader for ElementStreamReader {
     }
 
     fn read_bool(&mut self) -> IonResult<bool> {
-        self.current_value
-            .as_ref()
-            .and_then(|v| v.as_bool())
-            .ok_or_else(|| self.expected("bool value"))
+        self.current_value_as("bool value", |v| v.as_bool())
     }
 
     fn read_integer(&mut self) -> IonResult<Integer> {
-        self.current_value
-            .as_ref()
-            .and_then(|v| v.as_integer())
-            .map(|v| v.to_owned())
-            .ok_or_else(|| self.expected("int value"))
+        self.current_value_as("int value", |v| v.as_integer().map(|i| i.to_owned()))
```

Comment:
Here and below, consider moving the `map` logic to after the error handling to be even more concise:

```suggestion
        self.current_value_as("int value", |v| v.as_integer())?.to_owned()
```

Reply:
This again results into lifetime issues coming from `Element#as_*()`.